### DISCUSSION TITLE
Windows clients and a VPN gateway with two subnets

### DISCUSSION THREAD

**Discussion by assen-totin** [strongswan/strongswan#1679] [General]
Hello!

We need to connect road warriors to a LAN that has two independent subnets, 172.16/16 and 10/8. We want to avoid having a default route via the VPN.

Out basic setup includes StrongSwan and FreeRadius that handles authentication against Windows AD via NTLM using MS-CHAPv2. On the StrongSwan side, in the connection we have a single child SA with two traffic selectors, 172.16.0.0/16 and 10.0.0.0/8. 

For Linux and MacOS clients, everything works as expected: both subnets are accessible and the default route is not on the VPN.

We're considering the case of having Windows clients (only Windows 10/11 has to be supported):
- Keep the "default route on the VPN" checkbox disabled in Windows.
- Uncheck the "install classful routes" checkbox too.
- Install on the client two routes that point 172.16/16 and 10/8 to the VPN. 

Could such routes be pushed by StrongSwan upon connecting? I know of DHCP option 121, but my impression is that "connections.<conn>.pools=dhcp" only makes StrongSwan request the IP address for the VPN client via DHCP - and then it's passed back to the client as IKEv2 parameter, not? I don't see any native equivalent to DCHP option 121 in IKEv2 parameters - am I wrong? Upon connecting the VPN, would the Windows client make a real DHCP request -- and would it be sent over the IPSec link? (If yes, then maybe we should not configure any pool in StrongSwan at all and let this "native" DHCP request reach a real DCHP server over the VPN - but how would then StrongSwan learn the client's IP address - and should it really know it?). 

Or is this all wishful-thinking and I should go back to the well-known Add-VpnConnectionRoute?

For the record, I've read https://github.com/strongswan/strongswan/discussions/1055 and it does not give me much hope, but, hey, might still be worth asking?

**Comment by tobiasbrunner** [comment]
> I know of DHCP option 121, but my impression is that "connections..pools=dhcp" only makes StrongSwan request the IP address for the VPN client via DHCP - and then it's passed back to the client as IKEv2 parameter, not?

Yes, that option has nothing to do with the [_dhcp_ plugin](https://docs.strongswan.org/docs/5.9/plugins/dhcp.html).




> I don't see any native equivalent to DCHP option 121 in IKEv2 parameters - am I wrong?

Traffic selectors for one (it's what the other clients use). And there is the `INTERNAL_IP4_SUBNET` configuration attribute that some clients seem to request for this. But I don't think Windows clients do so. (strongSwan currently doesn't really support that attribute anyway.)



> Upon connecting the VPN, would the Windows client make a real DHCP request -- and would it be sent over the IPSec link?

Could be, at least some [Windows Mobile clients did that at some point](https://docs.strongswan.org/docs/5.9/interop/windowsClients.html#_split_routing_since_windows_10) (I know the section title there doesn't mention "Mobile", that apparently got lost when the docs were migrated, but I don't know if non-Mobile Windows versions ever sent DHCP via tunnel).



> (If yes, then maybe we should not configure any pool in StrongSwan at all and let this "native" DHCP request reach a real DCHP server over the VPN - but how would then StrongSwan learn the client's IP address - and should it really know it?).

You'll still have to assign a virtual IP to the client from a pool as the client's traffic selector would otherwise not get narrowed. DHCP would only be used to send DHCP option 121/249.

**Reply by assen-totin** [reply]
Thank you, Tobias.

I'm still on my quest to see if it would be possible to inject the routes to Windows clients (instead of teaching them the voodoo of PowerShell).

It seems (as per a few articles scattered around the Internet) that Windows clients send a DHCPINFORM packet after connecting the VPN. It is reported to originate from the IP address that was given over IKEv2 (port 68) and to be targeting 255.255.255.255:67. (Note that this happens without the familiar DHCP discover/offer/request stanza before it.)

On the (testing) VPN server I have a very simple setup: two regular interfaces (Ethernet), one public, one private; Strongswan is on the public one. No firewall even, only forwarding via kernel sysctl. Still, I don't see such a broadcast packet on neither of the two interfaces.

Next, I stumbled onto the "forecast" plugin for Charon. Is it required to make a broadcast from the VPN client reach any of the network interfaces on the VPN gateway? (I have no intention to re-inject the broadcast packet into other tunnels; I only want to make it visible on the LAN interface, where a DHCP server could pick it up). Or is there something else I should do to have these broadcast packets from the VPN client reach a (regular Ethernet) interface on the VPN gateway?

Also, I see some remarks that the broadcast address should be included in the traffic selector. If I want the VPN client to be able to send a broadcast to 255.255.255.255 and capture it on the VPN gateway, should I include 255.255.255.255/32 in the left TS? (To remind, I do not want to have 0.0.0.0/0 there as I do not want to have a default route via the VPN).

Why do I ask, instead of just trying it? Because the RHEL/EPEL/Fedora maintainer has excluded this plugin (plus a few more) from the package with the rather vague remark saying "too broken to enable". If you say the forecast plugin is a must-have, I'll need to rebuild their source RPM.

Thank you in advance,

Assen

**Reply by tobiasbrunner** [reply]
> Next, I stumbled onto the "forecast" plugin for Charon. Is it required to make a broadcast from the VPN client reach any of the network interfaces on the VPN gateway?

No, the server should receive it either way.

> Also, I see some remarks that the broadcast address should be included in the traffic selector. If I want the VPN client to be able to send a broadcast to 255.255.255.255 and capture it on the VPN gateway, should I include 255.255.255.255/32 in the left TS?

Yes, even if Windows would send such a packet, it would get dropped on the server by the IPsec policies. So you have to either install such a policy manually or change the traffic selectors.

**Comment by assen-totin** [comment]
Thank you, Tobias.

For the record, it does work.
- Regular desktop Windows clients do indeed send a DHCPINFORM using their just-assigned IP address as source and broadcast destination address. 
- Adding 255.255.255.255/32 to traffic selector allows StrongSwan to pass this broadcast to the local LAN.
- The broadcast will only appear on the interface where the particular StrongSwan connection is listening. 

Some additional info for the posterity:
- In the DHCPINFORM, Windows clients set their hardware type to 0x08 and hardware address size to 0x00 - which most DHCP servers will not be able to handle (incl. FreeRADIUS) as they firmly expect Ethernet (0x01) and 6-byte addressing. If you want to respond, you probably need a custom DHCP server.
- Regarding requested routes in the DHCPINFORM, Windows will only ask for option 249 (and not for the standard one 121). You must respond to them with 249 as the client may ignore 121 since it has not asked for it.
